### ISSUE TITLE
Anonymous lifetimes in impl trait are allowed in async fn, despite being an unstable feature

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust#108468]
I tried this code ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1f5e61ad851b4da09ba552c4268ac16d)):

```rust
async fn test(args: impl Iterator<Item = &str>) {}
```

I expected to see this happen: an error because `&str` contains an anonymous lifetime, and is part of an `impl Trait` type.

Instead, this happened: Compiles successfully.

Removing `async` from the function definition gives the following error (on nightly to see the feature gate):

```
error[[E0658]](https://doc.rust-lang.org/nightly/error_codes/E0658.html): anonymous lifetimes in `impl Trait` are unstable
 --> src/lib.rs:1:37
  |
1 | fn test(args: impl Iterator<Item = &str>) {}
  |                                     ^ expected named lifetime parameter
  |
  = help: [add `#![feature(anonymous_lifetime_in_impl_trait)]`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=1f5e61ad851b4da09ba552c4268ac16d#) to the crate attributes to enable
help: consider introducing a named lifetime parameter
  |
1 | fn test<'a>(args: impl Iterator<Item = &'a str>) {}
  |        ++++                             ++
```

This feature seems to have no documentation or tracking issue, only an [unstable book entry](https://dev-doc.rust-lang.org/stable/unstable-book/language-features/anonymous-lifetime-in-impl-trait.html#anonymous_lifetime_in_impl_trait). Given this, and the fact this behavior has been stable on async accidentally (AFAICT) for many releases now, maybe the best solution will be to just un-featuregate this functionality.

### Meta

`rustc --version --verbose`:
```
rustc 1.67.1 (d5a82bbd2 2023-02-07)
```

**Comment by clubby789** [comment]
This is explicitly allowed

https://github.com/rust-lang/rust/blob/31448badfd74ea72d2c8622cc60d3dca889ef7d4/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs#L1210-L1211

**Comment by jplatte** [comment]
That comment says "**used to be** allowed in async fns and forbidden in regular fns" though?

**Comment by traviscross** [comment]
@rustbot labels +AsyncAwait-Triaged +T-lang

We discussed this in async triage.  This is pending action from lang, and we've renominated #107378.

**Comment by ijackson** [comment]
I just came across this. The note about #107378 seems out of date.  See #137575.

It would be nice if in the meantime the error message said something more true.  I've filed #149076 for that.
